near-sdk-macros/src/lib.rs

```diff
@@ -207,6 +209,174 @@ pub fn metadata(item: TokenStream) -> TokenStream {
+            [
+                (json_schema, &["serde", "schemars", "validate"][..]),
+                (borsh_schema, &["borsh_skip"][..]),
+            ]
+            .iter()
+            .any(|&(case, paths)| case && paths.iter().any(|path| attr.path.is_ident(path)))
+        });
+    };
+
+    match &mut input.data {
+        syn::Data::Struct(data) => {
+            for field in &mut data.fields {
+                strip_unknown_attr(&mut field.attrs);
+            }
+        }
+        syn::Data::Enum(data) => {
+            for variant in &mut data.variants {
+                strip_unknown_attr(&mut variant.attrs);
+                for field in &mut variant.fields {
+                    strip_unknown_attr(&mut field.attrs);
+                }
+            }
+        }
+        syn::Data::Union(_) => {
+            return TokenStream::from(
+                syn::Error::new(Span::call_site(), "Near schema does not support unions yet.")
+                    .to_compile_error(),
+            );
+        }
+    }
+
+    let derive = match (json_schema, borsh_schema) {
+        // <unspecified> or #[abi(json)]
+        (_, false) => quote! {
+            #[derive(schemars::JsonSchema)]
+        },
+        // #[abi(borsh)]
+        (false, true) => quote! {
+            #[derive(borsh::BorshSchema)]
+        },
+        // #[abi(json, borsh)]
+        (true, true) => quote! {
+            #[derive(schemars::JsonSchema, borsh::BorshSchema)]
+        },
+    };
+
+    let input_ident = &input.ident;
+
+    let json_impl = if json_schema {
+        quote! {
+            #[automatically_derived]
+            impl schemars::JsonSchema for super::#input_ident {
+                fn schema_name() -> ::std::string::String {
+                    stringify!(#input_ident).to_string()
+                }
+
+                fn json_schema(gen: &mut schemars::gen::SchemaGenerator) -> schemars::schema::Schema {
+                    <#input_ident as schemars::JsonSchema>::json_schema(gen)
+                }
+            }
+        }
+    } else {
+        quote! {}
+    };
+
+    let borsh_impl = if borsh_schema {
+        quote! {
+            #[automatically_derived]
+            impl borsh::BorshSchema for super::#input_ident {
+                fn declaration() -> ::std::string::String {
+                    stringify!(#input_ident).to_string()
+                }
+
+                fn add_definitions_recursively(
+                    definitions: &mut borsh::maybestd::collections::HashMap<
+                        borsh::schema::Declaration,
+                        borsh::schema::Definition,
+                    >,
+                ) {
+                    <#input_ident as borsh::BorshSchema>::add_definitions_recursively(definitions);
+                }
+            }
+        }
+    } else {
+        quote! {}
+    };
+
+    TokenStream::from(quote! {
+        #[cfg(not(target_arch = "wasm32"))]
+        const _: () = {
+            mod __near_abi_private {
```

Comment:
`NearSchema`'s expansion effectively adds derives to a clone of the structure.

```rust
#[derive(NearSchema)]
struct Value {
    field: InnerValue
}
```

expands into

```rust
struct Value {
    field: InnerValue
}

const _: () = {
    mod __near_abi_private {
        #[derive(schemars::JsonSchema)]
        struct Value {
            field: InnerValue
        }
        
        impl schemars::JsonSchema for super::Value { ... }
        //                            ^^^^^
        // disambiguates the cloned structure from the original one
    }
};
```

Reply:
oh, I see, so you want to avoid the type name being something different from the original for the purposes of schemars, which uses the type name or something?

Reply:
Yeah, borsh also uses the type name.

Reply:
Bump on this. `super::*` requires the type to exist in a module. And would fail in this case.

```rust
mod x {
    fn a() {
        #[derive(NearSchema)]
        pub struct MyType;
    }
}
```

`super::MyType` from the injected `__near_abi_private` module would reference `x::MyType`, which is nonexistent.

Reply:
Resolved to expanding to this instead:

```rust
struct Value {
    field: InnerValue
}

const _: () = {
    type Proxy = Value;
    {
        #[derive(schemars::JsonSchema)]
        struct Value {
            field: InnerValue
        }
        
        impl schemars::JsonSchema for Proxy { ... }
    }
};
```